### PR TITLE
fix(logger-scope): resolve NullPointerException in MUnit tests

### PR DESCRIPTION
**Pull request by sachin-vcs**
- Make ConfigsSingleton configs map static final ConcurrentHashMap so all DI instances share the same config entries across MUnit's separate classloading/lifecycle context
- Add removeConfig() to ConfigsSingleton and call it from JsonloggerConfiguration.dispose() to prevent stale entries in the now-shared static map across redeploy/test cycles
- Inject Mule Registry into JsonloggerOperations and use Registry.lookupByName() as a fallback when ConfigsSingleton returns null, triggering MUnit's lazy config initialization which populates the singleton via initialise() as a side effect
- Re-check ConfigsSingleton after the Registry lookup to retrieve the now-populated config
- Add graceful degradation with ERROR-level logging when config still cannot be resolved, executing inner chain processors without JSON logging instead of throwing NPE
- Replace all repeated configs.getConfig(configurationRef) calls in loggerScope with a single resolved local variable to reduce null exposure and avoid dispose/in-flight race conditions
- Refactor publishScopeLogEvents to accept the resolved config object instead of re-reading from ConfigsSingleton by name

### CODE REVIEW

**Comment by mtransier** [comment]
Hi @sachin-vcs,
Thank you for your PR. Can you please elaborate a bit on what issue it is fixing?

File: src/main/java/org/mule/extension/jsonlogger/internal/JsonloggerOperations.java [hunk 1]
```diff
@@ -392,10 +424,10 @@ public void loggerScope(@DisplayName("Module configuration") @Example("JSON_Logg
         }
     }
 
-    private void publishScopeLogEvents(String configurationRef, String correlationId, String finalLogBefore, String finalLogAfter) {
-        LOGGER.debug("externalDestination.getDestination().getSupportedCategories().isEmpty(): " + configs.getConfig(configurationRef).getExternalDestination().getSupportedCategories().isEmpty());
-        LOGGER.debug("externalDestination.getDestination().getSupportedCategories().contains(jsonLogger.getName()): " + configs.getConfig(configurationRef).getExternalDestination().getSupportedCategories().contains(jsonLogger.getName()));
-        if (configs.getConfig(configurationRef).getExternalDestination().getSupportedCategories().isEmpty() || configs.getConfig(configurationRef).getExternalDestination().getSupportedCategories().contains(jsonLogger.getName())) {
+    private void publishScopeLogEvents(JsonloggerConfiguration config, String configurationRef, String correlationId, String finalLogBefore, String finalLogAfter) {
```
Comment (mtransier) [inline src/main/java/org/mule/extension/jsonlogger/internal/JsonloggerOperations.java:427]: Why is this using config and configurationRef at the same time?

**Comment by sachin-vcs** [comment]
Hi @mtransier,

Please find the details of the fix below:

> Can you please elaborate a bit on what issue it is fixing?

This PR fixes the `NullPointerException` that occurs when running MUnit tests against flows that use `json-logger:logger-scope`. A similar issue was reported in [#37](https://github.com/mulesoft-consulting/json-logger/issues/37).

**Root cause:** The Mule SDK does not support `@Config` injection in scopes, so the `loggerScope` method relies on `ConfigsSingleton` (populated during `initialise()` in `JsonloggerConfiguration`) to resolve the config. In MUnit, extension configurations are lazily initialized. This means `loggerScope` executes *before* `initialise()` has been called, so the singleton is empty and `getConfig()` returns `null`, which causes the NPE.

**What the fix does:**

1. **Static shared config map:** The configs map in `ConfigsSingleton` is now a `static final ConcurrentHashMap` so config entries are shared across all DI-injected instances. This addresses a potential instance mismatch in MUnit's separate lifecycle context.
2. **Registry fallback for lazy init:** When the singleton returns `null`, `Registry.lookupByName(configurationRef)` is used as a fallback. This triggers MUnit's lazy initialization of the extension config, which calls `initialise()` and populates the singleton as a side effect. A re-check then retrieves the now-available config.
3. **Graceful degradation:** A last-resort path was added (ERROR log + execute inner chain without JSON logging) instead of NPE, for robustness in edge cases where config truly cannot be resolved.
4. **Dispose cleanup:** A `removeConfig()` method was added and is called from `dispose()` to clean up the now-shared static map across redeploy/test cycles.
5. **Local variable refactor:** Repeated `configs.getConfig(configurationRef)` calls in `loggerScope` were replaced with a single resolved local variable to reduce null exposure and avoid a dispose/in-flight race condition.

**Verified behavior with a Mule 4 application using `json-logger:logger-scope`:**

Before fix (MUnit):
```
ERROR ... java.lang.NullPointerException
  at o.m.e.jsonlogger.internal.JsonloggerOperations.loggerScope(JsonloggerOperations.java:277)
```

Error details when running debugging MUnit:

```
org.mule.runtime.core.internal.message.ErrorBuilder$DeserializableErrorImplementation
{
  description=Cannot invoke "org.mule.extension.jsonlogger.internal.JsonloggerConfiguration.getJsonOutput()" because the return value of "org.mule.extension.jsonlogger.internal.singleton.ConfigsSingleton.getConfig(String)" is null
  detailedDescription=Cannot invoke "org.mule.extension.jsonlogger.internal.JsonloggerConfiguration.getJsonOutput()" because the return value of "org.mule.extension.jsonlogger.internal.singleton.ConfigsSingleton.getConfig(String)" is null
  errorType=MULE:UNKNOWN
  cause=java.lang.NullPointerException
  errorMessage=-
  suppressedErrors=[]
  childErrors=[]
}
```

After fix (MUnit):
```
INFO  ... o.m.e.jsonlogger.JsonLogger: {"correlationId":"...","tracePoint":"OUTBOUND_REQUEST_SCOPE_BEFORE",...}
INFO  ... o.m.e.jsonlogger.JsonLogger: {"correlationId":"...","tracePoint":"OUTBOUND_REQUEST_SCOPE_AFTER",...}
...
Tests run: 1 - Failed: 0 - Errors: 0 - Skipped: 0
```

Full JSON Logger BEFORE/AFTER scope logging now works correctly during MUnit execution, with no warnings or errors.

---

> Why is this using config and configurationRef at the same time?

`config` is the resolved `JsonloggerConfiguration` object. It is used here to access `getExternalDestination()` and `getSupportedCategories()` without re-reading from the singleton. This avoids a potential dispose/in-flight race now that `dispose()` removes entries from the shared static map.

`configurationRef` (the String config name) is still needed because `publishToExternalDestination()` in `LogEventSingleton` takes the config name as a String parameter. That API is not being changed in this PR, so both values need to be passed through.

Please let me know in case of any further queries/clarifications.

**Comment by sachin-vcs** [comment]
Hi @mtransier
Just checking if you had a chance to review the details above. Happy to address any further questions/clarifications.

**Comment by mtransier** [comment]
Hi Sachin, not yet. I have it on my list.

**Comment by mtransier** [comment]
Hi @sachin-vcs ,
Can you please have a look at [PR109](https://github.com/anypointcloud/json-logger/pull/109)? I should solve the issue in a slightly different way which I like better. Does it work for you?

**Comment by sachin-vcs** [comment]
Hi @mtransier,

I've gone through the issue #109 & the new PR #110. The push-based destination registration in `LogEventSingleton` is a good change. Removing the eager `configs.getConfigs().forEach(...)` from `LogEventSingleton.initialise()` gets rid of that startup race condition.

I built and tested source branch of the PR #110 against a Mule 4 app that uses `json-logger:logger-scope` in MUnit, and the `loggerScope` NPE still occurs:

```
ERROR ... java.lang.NullPointerException: Cannot invoke "...JsonloggerConfiguration.getJsonOutput()"
  because the return value of "...ConfigsSingleton.getConfig(String)" is null
  at ...JsonloggerOperations.loggerScope(JsonloggerOperations.java:309)
Tests run: 1 - Failed: 0 - Errors: 1 - Skipped: 0
BUILD FAILURE
```

It looks like the two PRs are fixing different parts of the problem:

- **PR #110** fixes the race in `LogEventSingleton` where external destinations were loaded before configs were registered.
- **PR #108** fixes the NPE in `loggerScope` itself, where `configs.getConfig(configurationRef)` returns null because `ConfigsSingleton` hasn't been populated yet during MUnit's lazy initialization.

PR #110 doesn't change `ConfigsSingleton` or `JsonloggerOperations`, so the `loggerScope` code path is unaffected. `configs.getConfig(configurationRef)` is still called without any null protection, and the singleton map is still an instance-level `HashMap` that can be empty when MUnit's DI context differs.

The `logger-scope` MUnit test added in PR #110 should also reproduces this. Would it make sense to combine the relevant parts of both PRs? The `LogEventSingleton` changes from PR #110 plus the `ConfigsSingleton`/`loggerScope` fixes from PR #108 would cover both code paths.

I can incorporate the `LogEventSingleton` changes from PR #110 into PR #108 if that works, or you could add the `ConfigsSingleton`/`loggerScope` fixes as another commit on your branch. Either way works for me.

**Comment by mtransier** [comment]
Hi @sachin-vcs ,
I guess I still haven't understood your scenario. Are you testing a flow that uses the logger scope or are you logging within your MUnit test? Why would you do the latter? Can you maybe post a screenshot of your flow setup?

**Comment by sachin-vcs** [comment]
Hi @mtransier,

PFB the simple flow I've created:
```xml
<?xml version="1.0" encoding="UTF-8"?>

<mule
	xmlns:json-logger="http://www.mulesoft.org/schema/mule/json-logger"
	xmlns="http://www.mulesoft.org/schema/mule/core"
	xmlns:doc="http://www.mulesoft.org/schema/mule/documentation"
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xsi:schemaLocation="http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
http://www.mulesoft.org/schema/mule/json-logger http://www.mulesoft.org/schema/mule/json-logger/current/mule-json-logger.xsd">
	<json-logger:config
		name="JSON_Logger_Config"
		doc:name="JSON Logger Config"
		doc:id="14c9f26c-2f07-4ef2-bee8-1d1b7a42ae2e"
		applicationName="logging"
		applicationVersion="1.0.0"
		environment="sandbox" />
	<flow
		name="loggingFlow"
		doc:id="03e268d9-952c-4cfc-aaea-34f48bb059d4">
		<scheduler
			doc:name="Every 60s with 5s delay"
			doc:id="496e2395-ae46-496b-85d0-6c649c876612">
			<scheduling-strategy>
				<fixed-frequency
					frequency="60"
					startDelay="5"
					timeUnit="SECONDS" />
			</scheduling-strategy>
		</scheduler>
		<logger
			level="INFO"
			doc:name="INFO: Init"
			doc:id="396b7349-d307-4307-b017-1cefba686b49"
			message="Init log" />
		<json-logger:logger-scope
			doc:name="Logger scope"
			doc:id="0187607a-e88b-4a31-9938-1d3d4a9ce261"
			configurationRef="JSON_Logger_Config">
			<set-payload
				value="#[%dw 2.0&#10;output application/json&#10;---&#10;{&#10;	message: 'Hello world!!!'&#10;}]"
				doc:name="Hello World"
				doc:id="0f4c9d61-fc83-481f-a52d-6c518e0cbb41" />
		</json-logger:logger-scope>
		<logger
			level="INFO"
			doc:name="INFO: Exit"
			doc:id="80d040d9-d92f-48ab-bbc9-fdcad5da30d1"
			message="Exit log" />
	</flow>
</mule>
```

...and the associated MUnit test I am trying to run:

```xml
<?xml version="1.0" encoding="UTF-8"?>

<mule
	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
	xmlns:munit="http://www.mulesoft.org/schema/mule/munit"
	xmlns:munit-tools="http://www.mulesoft.org/schema/mule/munit-tools"
	xmlns="http://www.mulesoft.org/schema/mule/core"
	xmlns:doc="http://www.mulesoft.org/schema/mule/documentation"
	xsi:schemaLocation="
		http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
		http://www.mulesoft.org/schema/mule/munit http://www.mulesoft.org/schema/mule/munit/current/mule-munit.xsd
		http://www.mulesoft.org/schema/mule/munit-tools  http://www.mulesoft.org/schema/mule/munit-tools/current/mule-munit-tools.xsd">
	<munit:config name="logging-test-suite.xml" />
	<munit:test
		name="logging-test-suite-loggingFlowTest"
		doc:id="91dd32cb-24a5-4cb7-ba74-7768c466b03a"
		description="Test">
		<munit:execution>
			<flow-ref
				doc:name="Flow-ref to loggingFlow"
				doc:id="5894f404-2c33-458e-a952-04f095c71dc7"
				name="loggingFlow" />
		</munit:execution>
		<munit:validation>
			<logger
				level="INFO"
				doc:name="INFO: MUnit run succeeded"
				doc:id="34daf4af-cd08-45f4-9272-1c8d3b113806"
				message="MUnit run succeeded" />
		</munit:validation>
	</munit:test>
</mule>
```

When running the above unit test using the current release or the built version from [109 branch](https://github.com/anypointcloud/json-logger/tree/109-nullpointerexception-occurs-in-logger-scope-during-munit-tests), the previously mentioned NPE occurs.

Please let me know for any further questions/clarifications.

**Comment by mtransier** [comment]
Hi @sachin-vcs ,
Thank you for the example, that clarifies the actual issue. I had indeed fixed a different one involving external destinations. I guess yours can be pinned down to testing a flow with only a Logger Scope, while also no other flow in the current scope contains a JSON Logger operation. In these cases, the config is not automatically initialized as it is only referenced by name. I have taken your valuable input and created a minimal change, which uses the same side effect to instantiate the config. Please have a look at the updated PR.

**Comment by sachin-vcs** [comment]
Hi @mtransier,

Thanks for the update. I tested the latest version of your branch and confirmed the `loggerScope` NPE is resolved now.

One thing worth noting: in your version, the `Registry.lookupByName()` fallback lives inside `ConfigsSingleton.getConfig()`, and `loggerScope` calls `configs.getConfig(configurationRef)` multiple times during a single execution (for `getCachedTimerTimestamp`, `getJsonOutput`, `getExternalDestination`, etc.). In the normal case this is fine since the first call populates the map and subsequent calls hit the cache. But a few things from PR #108 might still be worth considering for a future commit:

- **Thread safety:** The `configs` map is still a plain `HashMap`. If multiple threads call `getConfig()` and `addConfig()` at the same time, that's not safe. PR #108 used `ConcurrentHashMap` for this reason.
- **Dispose cleanup:** There's no `removeConfig()` call during `dispose()`, so stale config entries can linger in the map across test cycles or hot redeploys.
- **Null guard:** If `Registry.lookupByName()` doesn't populate the config for whatever reason, `getConfig()` returns `null` and `loggerScope` will still NPE. PR #108 had a fallback that logs an ERROR and runs the inner chain without JSON logging instead of crashing.

None of these are blockers for merging. Just wanted to flag them for awareness. Happy to close this PR if you'd prefer to go with #110.

**Comment by mtransier** [comment]
Hi @sachin-vcs,

> One thing worth noting: in your version, the `Registry.lookupByName()` fallback lives inside `ConfigsSingleton.getConfig()`, and `loggerScope` calls `configs.getConfig(configurationRef)` multiple times during a single execution (for `getCachedTimerTimestamp`, `getJsonOutput`, `getExternalDestination`, etc.). In the normal case this is fine since the first call populates the map and subsequent calls hit the cache.

Exactly, that's the rationale behind this change.

> * **Thread safety:** The `configs` map is still a plain `HashMap`. If multiple threads call `getConfig()` and `addConfig()` at the same time, that's not safe. PR [fix(logger-scope): resolve NullPointerException in MUnit tests #108](https://github.com/anypointcloud/json-logger/pull/108) used `ConcurrentHashMap` for this reason.

I've updated the implementation accordingly.

> * **Dispose cleanup:** There's no `removeConfig()` call during `dispose()`, so stale config entries can linger in the map across test cycles or hot redeploys.

I've added this.

> * **Null guard:** If `Registry.lookupByName()` doesn't populate the config for whatever reason, `getConfig()` returns `null` and `loggerScope` will still NPE. PR [fix(logger-scope): resolve NullPointerException in MUnit tests #108](https://github.com/anypointcloud/json-logger/pull/108) had a fallback that logs an ERROR and runs the inner chain without JSON logging instead of crashing.

I actively decided to solve it like this, as the main reason would probably be that a scope references a non-existing configuration, in which case it would be better to run in to an error instead of gracefully ignoring the scope.

I will go ahead and merge the PR #110.

Thanks again for your input!

Best regards,
Matthias
